packages/ember-routing/lib/services/routing.js

```diff
@@ -0,0 +1,102 @@
+/**
+@module ember
+@submodule ember-routing
+*/
+
+import Service from "ember-runtime/system/service";
+
+import { get } from "ember-metal/property_get";
+import { readOnly } from "ember-metal/computed_macros";
+import { routeArgs } from "ember-routing/utils";
+import keys from "ember-metal/keys";
+import merge from "ember-metal/merge";
+
+/** @private
+  The Routing service is used by LinkView, and provides facilities for
+  the component/view layer to interact with the router.
+
+  While still private, this service can eventually be opened up, and provides
+  the set of API needed for components to control routing without interacting
+  with router internals.
+*/
+
+var RoutingService = Service.extend({
+  router: null,
+
+  targetState: readOnly('router.targetState'),
+  currentState: readOnly('router.currentState'),
+  currentRouteName: readOnly('router.currentRouteName'),
+
+  availableRoutes: function() {
+    return keys(get(this, 'router').router.recognizer.names);
+  },
+
+  hasRoute: function(routeName) {
+    return get(this, 'router').hasRoute(routeName);
+  },
+
+  transitionTo: function(routeName, models, queryParams, shouldReplace) {
+    var router = get(this, 'router');
+
+    var transition = router._doTransition(routeName, models, queryParams);
+
+    if (shouldReplace) {
+      transition.method('replace');
+    }
+  },
+
+  normalizeQueryParams: function(routeName, models, queryParams) {
+    get(this, 'router')._prepareQueryParams(routeName, models, queryParams);
+  },
+
+  generateURL: function(routeName, models, queryParams) {
+    var router = get(this, 'router');
+
+    var visibleQueryParams = {};
+    merge(visibleQueryParams, queryParams);
+
+    this.normalizeQueryParams(routeName, models, visibleQueryParams);
+
+    var args = routeArgs(routeName, models, visibleQueryParams);
+    return router.generate.apply(router, args);
+  },
+
+  isActiveForRoute: function(params, routeName, routerState, isCurrentWhenSpecified) {
+    var router = get(this, 'router');
+    var contexts = params.models;
+
+    var handlers = router.router.recognizer.handlersFor(routeName);
+    var leafName = handlers[handlers.length-1].handler;
+    var maximumContexts = numberOfContextsAcceptedByHandler(routeName, handlers);
```

Comment:
i wonder if we can cache this, as long lists of link-to's get hit kinda hard right now.